conda/resolve.py

```diff
@@ -143,481 +164,429 @@ def __ge__(self, other):
+        return (VersionOrder(rec['version']), len(self.features(fn)), rec['build_number'], rec['build'])
 
     @memoize
     def features(self, fn):
+        if fn[-1] == ']':
+            return self.features(fn.rsplit('[',1)[0])
         return set(self.index[fn].get('features', '').split())
 
     @memoize
     def track_features(self, fn):
+        if fn[-1] == ']':
+            return self.track_features(fn.rsplit('[',1)[0])
         return set(self.index[fn].get('track_features', '').split())
 
-    # Changed to manual memoization so that we can reset if we add/change the filter.
-    def get_pkgs(self, ms, max_only=False):
-        key = (ms,max_only)
-        if key in self.get_pkgs_:
-            return self.get_pkgs_[key]
-        pkgs = [Package(fn, self.index[fn]) for fn in self.find_matches(ms)]
-        if not pkgs:
-            raise NoPackagesFound("No packages found in current %s channels matching: %s" % (config.subdir, ms), [ms.spec])
-        if max_only:
-            maxpkg = max(pkgs)
-            ret = []
-            for pkg in pkgs:
-                try:
-                    if (pkg.name, pkg.norm_version, pkg.build_number) == \
-                       (maxpkg.name, maxpkg.norm_version, maxpkg.build_number):
-                        ret.append(pkg)
-                except TypeError:
-                    # They are not equal
-                    pass
-            pkgs = ret
-        self.get_pkgs_[key] = pkgs
-        return pkgs
-
-    def get_max_dists(self, ms):
-        pkgs = self.get_pkgs(ms, max_only=True)
-        if not pkgs:
-            raise NoPackagesFound("No packages found in current %s channels matching: %s" % (config.subdir, ms), [ms.spec])
-        for pkg in pkgs:
-            yield pkg.fn
-
-    def all_deps(self, root_fn, max_only=False):
-        res = {}
-
-        def add_dependents(fn1, max_only=False):
-            for ms in self.ms_depends(fn1):
-                found = False
-                notfound = []
-                for pkg2 in self.get_pkgs(ms, max_only=max_only):
-                    if pkg2.fn in res:
-                        found = True
-                        continue
-                    res[pkg2.fn] = pkg2
-                    try:
-                        add_dependents(pkg2.fn, max_only=max_only)
-                    except NoPackagesFound as e:
-                        for pkg in e.pkgs:
-                            if pkg not in notfound:
-                                notfound.append(pkg)
-                        if pkg2.fn in res:
-                            del res[pkg2.fn]
-                    else:
-                        found = True
+    @memoize
+    def package_triple(self, fn):
+        if not fn.endswith('.tar.bz2'):
+            return self.package_triple(fn + '.tar.bz2')
+        rec = self.index.get(fn, None)
+        if rec is not None:
+            return (rec['name'], rec['version'], rec['build'])
+        return fn[:-8].rsplit('-',2)[0]
```

Comment:
Interesting. Can you give me an example where this is actually occurring? I'm not saying we shouldn't fix it, but I wonder if it should be fixed upstream, and maybe place an `assert` here.